src/node_api.cc

```diff
@@ -918,21 +919,26 @@ NAPI_NO_RETURN void napi_fatal_error(const char* location,
                                      size_t location_len,
                                      const char* message,
                                      size_t message_len) {
-  char* location_string = const_cast<char*>(location);
-  char* message_string = const_cast<char*>(message);
-  if (location_len != -1) {
-    location_string = reinterpret_cast<char*>(
-        malloc(location_len * sizeof(char) + 1));
-    strncpy(location_string, location, location_len);
-    location_string[location_len] = '\0';
-  }
-  if (message_len != -1) {
-    message_string = reinterpret_cast<char*>(
-        malloc(message_len * sizeof(char) + 1));
-    strncpy(message_string, message, message_len);
-    message_string[message_len] = '\0';
-  }
-  node::FatalError(location_string, message_string);
+  std::string location_string;
+  std::string message_string;
+
+  if (static_cast<int>(location_len) != NAPI_AUTO_LENGTH) {
```

Comment:
I don't think that `size_t` always fits into an `int`. Why do we need to cast here?

Reply:
I agree it should not be needed in this check.  In places where it is passed on to v8 then it is needed.

Reply:
@mhdawson No, the issue isn’t that it’s in the wrong place – it’s that a downcast is a *real* bug here, because it truncates `location_len`

Reply:
@addaleax, to clarify we know that size_t does not fit into an int, so where its going to be passed to v8 a cast is required to avoid warnings.  In this case we should just be comparing NAPI_AUTO_LENGTH directly against location_len.

Reply:
@mhdawson Yeah, but we can’t just downcast without any checks in this case. Imagine a platform that has 32-bit `int`, and 64-bit `size_t`. When an addon tries to create a N-API string that contains (2^32 + 1) characters – not knowing that that doesn’t work, because, how would it tell? – N-API would currently silently truncate that size argument to 1, and happily create a one-character string. I think that qualifies as a bug.